### PR TITLE
Docs: add content from old documentation on caching/hashing

### LINKED ISSUE #4472: Docs: Move docs on caching
From #4451 

Move the content of "Controlling hashing" and "Design guidelines" inside of `core/caching.rst` to `topics/provenance/caching` (see the [old documentation](https://aiida-core.readthedocs.io/en/v1.1.0/developer_guide/core/caching.html)).

### PR DESCRIPTION
**Pull request by ramirezfranciscof**
close #4472 (now for real)

Moved the content of "Controlling hashing" and "Design guidelines" inside of `developer_guide/core/caching.rst` to `topics/provenance/caching`.

### CODE REVIEW

**Comment by ramirezfranciscof** [comment]
@ltalirz I'm tagging you since we discussed this earlier so I think you have this fresh and it would be quick for you to review it, but let me know if you prefer for me to find someone else.

**Comment by ramirezfranciscof** [comment]
@ltalirz @sphuber this should be a fairly short change, mostly moved content around with minor adaptations. Could any of you take a look at it?

File: docs/source/topics/provenance/caching.rst
Comment (sphuber) [inline docs/source/topics/provenance/caching.rst:75]: This section should be clarified. Part of this is aimed at `aiida-core` developers and some of it can also be directed to plugin developers. The behavior that is described can only be adapted outside of `aiida-core` for `Data` subclasses, because process nodes cannot be subclassed. So I think this content should go in some form to the Internal Architecture section, but it should also stay here, in rewritten form, where it makes clear that this can only be done for `Data` plugins. With that given, it should also give more clear examples making it more practically targeted at plugin developers.

**Review by sphuber — CHANGES_REQUESTED** [review]
Thanks @ramirezfranciscof . If I understand correctly, you are moving in two sections essentially verbatim and adding a "todo" paragraph in the internal architecture. Besides not moving in all existing documentation, why not actually write the documentation instead of writing a small todo?

File: docs/source/topics/provenance/caching.rst
Comment (greschd) [reply docs/source/topics/provenance/caching.rst:75]: In the part for plugin developers, there should also be a mention of the `invalidates_cache` keyword argument on exit codes.

File: docs/source/internals/engine.rst
Comment (ramirezfranciscof) [reply docs/source/internals/engine.rst:16]: The engine part was being done by @chrisjsewell and a reminder to do this was already added to that issue. I added this here mostly because since we already had the "duplicated" Engine task (meaning, both the issue and the todo in the files), we might as well have all the info in both places. If I copy it myself it might conflict with the other content of the section and be annoying to resolve, not to mention that it might already have been included in that work.

File: docs/source/developer_guide/core/caching.rst
Comment (ramirezfranciscof) [reply docs/source/developer_guide/core/caching.rst:9]: [I understood](https://github.com/aiidateam/aiida-core/issues/4451#issuecomment-711792600) that the whole "Controlling Hashing" sections was supposed to go to the internal architecture page, not to the topics. Is it only the "Workflownode example" that has to go there then? If the main part of "Controlling Hashing" has user-facing relevant information, then the example also looks to be dealing with the same kind of stuff.

File: docs/source/internals/engine.rst
Comment (sphuber) [reply docs/source/internals/engine.rst:16]: If he is writing and improving the caching documentation, wouldn't it then make sense to have him also move/incorporate these parts, so he can make sure the whole thing is coherent?

Comment (ramirezfranciscof) [reply docs/source/internals/engine.rst:16]: Not the caching in particular, he is writing essentially from scratch the docs for the internals of the engine, which were lower priority than trying to get all of the old documentation back somewhere in the new and hence he didn't need to finish this for the release. What we did want was to try to get back as much of the old docs as possible, and since the "topics" section was already present, moving all of the "user-facing" content of caching there seemed like a possibility. If you think is better to wait to emphasize on the consistency over having available these sections on caching/hashing now for the release, then I guess I can block this until the engine section is finished.

File: docs/source/topics/provenance/caching.rst
Comment (ramirezfranciscof) [reply docs/source/topics/provenance/caching.rst:75]: > The behavior that is described can only be adapted outside of aiida-core for Data subclasses, because process nodes cannot be subclassed.

@sphuber These are all for `Data` subclasses? Then how do you use the `_hash_ignored_inputs`? Isn't there a way to control the hashing of processes via the definition of the calcjob plugin?

> In the part for plugin developers, there should also be a mention of the `invalidates_cache` keyword argument on exit codes.

@greschd Could you explain to me how does this work? I am not familiar with it.

Comment (sphuber) [reply docs/source/topics/provenance/caching.rst:76]: > sn't there a way to control the hashing of processes via the definition of the calcjob plugin?

There are some ways to control hashing of processes, for example the `invalidates_cache` of the `ExitCode`, but anything that is defined on the `Node` class can not be changed because node classes can not be subclassed and controlled by plugins

Comment (greschd) [reply docs/source/topics/provenance/caching.rst:76]: >  Could you explain to me how does this work? I am not familiar with it.

The `spec.exit_code` function takes an `invalidates_cache` keyword argument. If `True`, a calculation that has this exit code will not be used as a cache source.
Plugin developers can use this to e.g. disable caching for calculations that failed with an unknown error (that could well be node failure or similar). Simple example from `aiida-wannier90`: https://github.com/aiidateam/aiida-wannier90/pull/104

File: docs/source/internals/engine.rst
Comment (greschd) [inline docs/source/internals/engine.rst:30]: Since this is a way for plugin developers to override the caching behavior, shouldn't this be mentioned in the section for plugin developers?

File: docs/source/topics/provenance/caching.rst
Comment (greschd) [inline docs/source/topics/provenance/caching.rst:81]: This isn't controlling _hashing_, it's controlling _caching_. Since the title doesn't apply, maybe it should have a separate short section, together with the `is_valid_cache` mechanism?

**Review by greschd — COMMENTED** [review]
Thanks @ramirezfranciscof, looks good overall.

Since I don't have a top-level view of the documentation structure I'll leave the main review up to @sphuber.

File: docs/source/internals/engine.rst
Comment (ramirezfranciscof) [reply docs/source/internals/engine.rst:30]: I interpret that you mean it in the sense that "...shouldn't this ALSO be mentioned...", right? Since here it is explaining the internal mechanisms by which the feature is implemented (in the same way the 3rd point below this explains where the `exit_codes` with `invalidate_cache` are processed). If that is the case, yes, I agree there should be also a more "application oriented" reference to this in the topics.

Comment (greschd) [reply docs/source/internals/engine.rst:30]: Yep, _also_ mentioned in the topics.

**Comment by ramirezfranciscof** [comment]
Thanks for the comments @greschd ! I have now incorporated them into the PR, let me know what you think.

**Comment by greschd** [comment]
Yep, all good from my POV. Still think someone else should give it a pass though, because I'm not familiar with the design of the docs.

**Comment by ramirezfranciscof** [comment]
Maybe @sphuber can give a last pass or suggest some other person to do this?

**Review by sphuber — APPROVED** [review]
Thanks @ramirezfranciscof . Let's merge this and then further improvements can be added later on

**Pull request merged**
